feat: Add TiMidity++ and FluidR3 for MIDI support

4Vertikal4 opened the pull request:
**Description:**

This PR aims to add MIDI playback support to the Heroic Games Launcher Flatpak by integrating TiMidity++ (version 2.15.0) and the FluidR3_GM soundfont. The goal is to enable MIDI music for games launched via Heroic, particularly older titles that rely on MIDI. I wanted to implement my[ local fix.](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/issues/4409)

**What has been successfully implemented:**

- TiMidity++ version 2.15.0 has been added as a new module:
It is configured with --enable-alsa and --enable-alsaseq to support ALSA output, including the ALSA sequencer interface.
- The FluidR3_GM soundfont is included as a separate module (fluidr3-soundfont).
- A series of patches have been created and applied to the TiMidity++ 2.15.0 source code. These patches address numerous "implicit function declaration" errors, allowing TiMidity++ to compile successfully with the modern compilers in the Freedesktop SDK 24.08. The patched files include au_a.c, freq.c, output.c, quantity.c, raw_a.c, timidity.c, and wave_a.c.
- An inline configuration file (custom-timidity.cfg) is created and installed to /app/share/timidity/timidity.cfg within the sandbox. This file directs TiMidity++ to use the included FluidR3_GM soundfont located at /app/share/soundfonts/.
- The Flatpak manifest (com.heroicgameslauncher.hgl.yml) builds successfully with these additions.

**Current Status & Known Issue:**

Despite successful compilation, packaging, and configuration file placement, TiMidity++ currently fails to initialize ALSA output devices at runtime when executed inside the Flatpak sandbox.

- The primary error observed when attempting to run timidity (e.g., timidity -iA -Os --verbose &, timidity -Ov --verbose &, or even just timidity some_midi_file.mid) inside the sandbox is: Couldn't open output device
- This is often preceded or accompanied by messages like Playmode 's' is not compiled in. (or similar, depending on the flags used), which appears to be a secondary symptom of the failure to open an audio device.
- This failure occurs even when attempting to direct output to the pulse ALSA device (timidity -Opulse --verbose &).
- The host system used for testing runs Fedora with PipeWire 1.4.2 managing audio. The Flatpak is built against the Freedesktop SDK 24.08.

**Further Investigation Needed:**

The current evidence suggests an issue with how the sandboxed TiMidity++ (using the SDK's alsa-lib) interacts with the host's PipeWire-emulated ALSA environment through the Flatpak ALSA portal. Further investigation is required to make TiMidity++ functional for MIDI output. Potential areas to explore include:

- Using tools like strace on the timidity process inside the sandbox to identify precisely which ALSA calls are failing or what resources it's unable to access.
- Exploring if different or additional configure flags for TiMidity++'s ALSA backend are necessary for better compatibility within this specific sandboxed PipeWire environment.
- Investigating if specific ALSA plugins, environment variables, or configurations are missing or misconfigured within the Flatpak sandbox that TiMidity++ might require.
- Considering if an alternative, perhaps more modern, MIDI synthesizer might offer better out-of-the-box compatibility with the Flatpak sandbox and PipeWire.

4Vertikal4 committed:
feat: Add TiMidity++ and FluidR3 for MIDI support

This commit introduces TiMidity++ (v2.15.0) and the FluidR3_GM
soundfont to enable MIDI playback for games.

Includes:
- TiMidity++ module compiled with ALSA and ALSA sequencer support.
- FluidR3_GM soundfont module.
- Patches for TiMidity++ 2.15.0 source to fix compilation errors
  with modern compilers (implicit function declarations).
- Configuration for TiMidity++ to use the FluidR3_GM soundfont.

Note: While compilation and packaging are successful, a runtime
issue prevents TiMidity++ from initializing ALSA output devices
within the Flatpak sandbox ('Couldn't open output device').
This requires further investigation as detailed in the upcoming PR.

Also, consolidated individual patch files into a single patch within a `patches/` directory for better organization.

flavioislima commented on timidity-au_a-includes.patch:
Can you move the patches to a patch folder?

4Vertikal4 replied:
Done!

4Vertikal4 committed:
-  Integrate TiMidity++ 2.15.0 with ALSA sequencer support
  -  Include FluidR3_GM soundfont for high-quality MIDI synthesis
  -  Copy ALSA configuration files and plugins for proper audio output
  -  Add PulseAudio client libraries and runtime dependencies
  -  Apply compilation patches for modern compiler compatibility
  -  Successfully tested with The Elder Scrolls: Arena

4Vertikal4 commented:
OK so far this PR successfully adds MIDI playback support to the Heroic Games Launcher Flatpak by integrating TiMidity++ 2.15.0 and the FluidR3_GM soundfont.

**What's Implemented:**

1. TiMidity++ 2.15.0 compiles and runs in Flatpak sandbox
2. ALSA sequencer interface fully functional
3. MIDI synthesis with FluidR3 soundfont
4. Audio output through ALSA/PipeWire system
5. Successfully tested with The Elder Scrolls: Arena

**Testing Results:**

TiMidity starting in ALSA server mode
Opening sequencer port: 130:0 130:1 130:2 130:3
Playing time: ~30 seconds
Notes cut: 0
Notes lost totally: 0

**Current Usage (Manual Setup):**

1. Start TiMidity++ in Flatpak

`flatpak run --command=bash com.heroicgameslauncher.hgl`
`/app/bin/timidity -iA &`

2. Launch Heroic in another terminal
`flatpak run com.heroicgameslauncher.hgl`

3. Play MIDI-enabled games. For example TES I Arena - it's a free game on GOG.

**Known Limitation:**
Currently requires manual TiMidity++ startup before launching MIDI games. This is not ideal for end-user experience. When I will have time I will work on integrate automatic MIDI synthesizer management into Heroic's launch logic (I hope it's possible).

### **To sum it up - what I'm trying to achieve:**

1. Enabling MIDI music in classic DOS and early Windows games
2. High-quality audio synthesis with professional soundfont
3. Foundation for fully automated MIDI support

4Vertikal4 committed:
Add automatic MIDI detection wrapper with process detection limitation

- Implement smart MIDI file detection in game directories
- Auto-start TiMidity++ for games with MIDI files (.mid, .midi)
- Check config files for MIDI references (MPU-401, SoundBlaster, etc.)
- Wrapper intercepts gamemoderun to enable transparent automation

Known limitation: Automatic game termination detection

- Game launches and MIDI works perfectly
- Manual 'Stop' button in Heroic works normally
- When user quits via in-game menu, Heroic doesn't auto-detect termination
- User must manually click 'Stop' in Heroic after quitting game

This is a process lifecycle detection issue, not a critical functionality break. Seeking community input on:

- Better interception points that preserve process detection
- How Heroic tracks game process termination
- Whether this limitation is acceptable vs manual approach

4Vertikal4 commented:
**Smart MIDI Detection Implementation**

**Automatic Detection Features:**

- Scans game directories for MIDI files (.mid, .midi)
- Detects MIDI configuration references in game config files
- Identifies games with MIDI/SoundBlaster references

**Process Management Challenge:**

- Wrapper intercepts gamemoderun in launch chain
- Game termination through in-game menus not auto-detected by Heroic
- Manual "Stop" button in Heroic works normally as fallback

**My questions:**

- Anyone know better integration points that won't break Heroic's process detection?
- How does Heroic actually track when games end? Understanding this could help find a cleaner solution.
- Should we accept this minor UX quirk, or stick with the manual approach for now?

**Current Status:** Fully functional MIDI support with transparent automation, minor process detection limitation requiring manual stop after in-game quit.

4Vertikal4 committed:
Add complete automatic MIDI detection with process tracking

- Implement MIDI file detection (.mid, .midi, config references)
- Auto-start TiMidity++ with FluidR3 soundfont for MIDI games
- Process wrapper preserves game lifecycle for proper Heroic integration
- Zero user configuration needed - works out of the box
- Tested successfully with The Elder Scrolls: Arena

4Vertikal4 commented:
> Nice, I will try to find some time to test it as well.

Any feedback? What should I correct?

flavioislima commented:
> > Nice, I will try to find some time to test it as well.
> 
> @flavioislima Any feedback? What should I correct?

I will try to test it this week. Or we can merge this in the beta branch and test there.

4Vertikal4 commented:
> > > Nice, I will try to find some time to test it as well.
> > 
> > 
> > @flavioislima Any feedback? What should I correct?
> 
> I will try to test it this week. Or we can merge this in the beta branch and test there.

@flavioislima I finished my other projects and wanted to look on status of this one. What is the progress of testing  TiMidity++ and FluidR3?

4Vertikal4 committed:
Update TiMidity++ build parameters for GCC 14 compatibility

4Vertikal4 commented:
@flavioislima  I've resurrected this PR and rebased it on top of the current master branch.

The Flatpak runtime has been updated significantly since my original implementation, so I also added the necessary build flags (-Wno-error=incompatible-pointer-types, etc.) and a quick sed patch to ensure the legacy TiMidity++ codebase compiles flawlessly with the modern GCC 14 compiler included in the latest Freedesktop SDK.

I have tested it locally with the latest Flatpak environment using The Elder Scrolls: Arena. The smart MIDI detection successfully injects the synthesizer, music plays perfectly through the sandboxed ALSA/PulseAudio routing, and automatic game exit detection works gracefully without requiring a manual "Stop".

Let me know if you need any further adjustments. The PR is fully green and ready for review!

flavioislima reviewed (approved):
Great stuff! Thanks!
